File: src/utils/errorMessage.ts
/**
 * Kullanıcıya gösterilecek hata metninin son kapısı.
 *
 * ## Neden var (13 Eylül 2026)
 *
 * Kullanıcı App Store için ekran kaydı çekerken giriş penceresinde şunu
 * gördü ve fotoğrafladı:
 *
 *     {"message":"Gateway Timeout"}
 *
 * Ham bir HTTP 504 gövdesi, Türkçe bir uygulamada, giriş formunun altında.
 * İkinci denemede giriş çalıştı — yani arıza geçiciydi, ama kullanıcının
 * gördüğü şey bunu SÖYLEMİYORDU.
 *
 * Bu tek bir unutulmuş satır değil, bir POLİTİKANIN sonucuydu.
 * `friendlyAuthMessage` (4 Ağustos 2026) ve `isNetworkError` ikisi de
 * bilerek *"eşleşmeyen hata HAM hâliyle geçsin"* diyordu; gerekçe de
 * sağlamdı: *"bilinmeyen bir hatayı uydurma bir Türkçe cümleyle gizlemek,
 * hata ayıklamayı imkânsız kılardı."*
 *
 * O gerekçe artık geçerli DEĞİL, çünkü arada `client_errors` telemetrisi
 * var (`utils/errorReporting.ts`, 30 Ağustos 2026). Ham metni kullanıcıya
 * BASMAK ile onu KAYBETMEK aynı şey değil: burası ham metni telemetriye
 * yazar, ekrana Türkçe bir cümle koyar. Hata ayıklama kabiliyeti artıyor —
 * çünkü eskiden yalnızca ekranı gören kişi biliyordu.
 *
 * ## Üç dal (repoda zaten vardı, burada tekleşti)
 *
 * Desenin kendisi yeni değil: `inviteAcceptErrorText`
 * (`mobile/app/lib/src/data/friends_api.dart`, 25 Ağustos 2026) tam bunu
 * yapıyordu, ama yalnızca TEK uçta. Genelleştirilmiş hâli:
 *
 * 1. **Sunucunun KENDİ reddi → olduğu gibi göster.** plpgsql'in
 *    `raise exception`'ı SQLSTATE **`P0001`** üretir ve bu projedeki o
 *    metinler (`submit_move` → "Sıra sende değil.", `accept_friend_invite`
 *    → "Kendi linkinle arkadaş olamazsın.") KULLANICIYA GÖSTERİLMEK üzere
 *    Türkçe yazılmıştır. Koda bakılır, METNE değil — metin değişebilir.
 * 2. **Geçici sunucu arızası → "birkaç saniye sonra tekrar dene".** 502/503/
 *    504, zaman aşımı, soket/el sıkışma hataları. Vakanın kendisi bu dal.
 * 3. **Makine metni → jenerik Türkçe cümle.** JSON/HTML gövdesi, PostgREST/
 *    GoTrue İngilizcesi, Dart/JS istisna dökümü, SQLSTATE dökümü.
 * 4. **Geri kalan → olduğu gibi.** Kendi Türkçe doğrulama mesajlarımız
 *    (`throw new Error('Ad zorunludur.')`) buradan geçer.
 *
 * ⚠ **Sıra davranışın parçası, üç yerde birden:**
 * · 1 her şeyden önce — `P0001` taşıyan bir mesaj makine kalıbına benzese
 *   bile gösterilir, sunucu onu bilerek yazdı.
 * · 2 mutlaka 3'ten önce — ekranda görülen `Gateway Timeout` düz METİN
 *   olarak da gelebiliyor (JSON gövdesi olmadan) ve o hâliyle hiçbir makine
 *   kalıbına takılmaz; sıra ters olsaydı 4. dala düşüp yine ham görünürdü.
 * · 2 ve 3 telemetriye YAZAR, 4 yazmaz — 4 zaten bizim metnimiz.
 *
 * ⚠ **`code` ancak taşınırsa işe yarar.** 13 Eylül 2026'ya kadar
 * `src/lib/api.ts` 45 yerde `throw new Error(error.message)` diyordu, yani
 * SQLSTATE yolda DÜŞÜYORDU ve 1. dal hiçbir zaman çalışamazdı. Aynı PR'da
 * hepsi `rethrowSupabase()`e çevrildi; yeni bir uç yazarken `code`'u
 * düşürme.
 *
 * ⚠ **Admin paneli bilerek DIŞARIDA.** `AdminDashboard`/`MemberMessageModal`
 * ham metni göstermeye devam eder: oranın tek kullanıcısı geliştiricinin
 * kendisi ve ham hata orada bir ARAÇ, arıza değil.
 *
 * Port ikizi: `mobile/app/lib/src/util/error_message.dart` — metinler ve
 * kalıplar BİREBİR aynı, `error_message_parity_test.dart` bu dosyayı okuyup
 * karşılaştırır.
 *
 * ⚠ **İkisi AYRI günlerde `main`'e girdi** (web 14 Eylül, port sonrası):
 * 1.1.0 (665) kodu donduruğu ve iki mağaza da incelediği için web yarısı
 * tek başına yayınlandı, port yarısı inceleme kapanana kadar bekletildi.
 * Yani `client_errors`ta 14 Eylül ile port sürümünün yayını arasındaki
 * pencerede `hata-metni:*` kayıtları YALNIZCA web'den gelir — mobil
 * sessizliği "mobilde hata yok" diye okunmamalı.
 */

/**
 * Ham metni nereye yazacağımız — DIŞARIDAN verilir.
 *
 * ⚠ Doğrudan `reportClientError` import EDİLMEZ (13 Eylül 2026'da denendi ve
 * geri alındı): o modül Supabase istemcisini çekiyor, istemci de
 * `import.meta.env` okuyor — yani bu dosyayı node'da import eden doğrulama
 * betiği (`npm run verify-error-messages`) daha ilk satırda düşüyordu.
 * Enjeksiyon hem saflığı hem kapıyı koruyor; deponun `ClientErrorSink`
 * deseninin aynısı.
 *
 * Bağlanmazsa sessizce hiçbir şey yazılmaz — metin kararı yine de doğrudur.
 */
type HataRaporlayici = (err: unknown, context: string) => void;
let raporla: HataRaporlayici | null = null;

/** `boot.tsx` açılışta bağlar (`installGlobalErrorReporting`'in yanında). */
export function setErrorMessageReporter(fn: HataRaporlayici | null): void {
  raporla = fn;
}

/** Bilinmeyen/makine kaynaklı hata — kullanıcı ne yapacağını bilsin. */
export const GENERIC_ERROR_NOTICE = 'Bir sorun oluştu. Lütfen tekrar dene.';

/**
 * Sunucu ayakta değil ya da zaman aşımına uğradı — GEÇİCİ olduğu söylenmeli.
 *
 * Ayrı bir metin olmasının sebebi doğrudan vakanın kendisi: kullanıcı ikinci
 * denemede giriş yapabildi. "Bir sorun oluştu" doğru ama eksik; burada
 * yapılacak somut bir şey var ve o da beklemek.
 */
export const TEMPORARY_ERROR_NOTICE =
  'Sunucuya şu anda ulaşılamıyor. Birkaç saniye sonra tekrar dene.';

/** Sunucunun bilerek yazdığı, kullanıcıya gösterilebilir ret (SQLSTATE P0001). */
export function isServerRejection(err: unknown): boolean {
  return (err as { code?: string } | null)?.code === 'P0001';
}

/**
 * Ağ geçidi / zaman aşımı / "sunucu ayakta değil" sınıfı.
 *
 * `isNetworkError`'dan (offlineNotice.ts) AYRI: orası "isteği hiç
 * gönderemedik" (cihaz çevrimdışı), burası "gönderdik ama sunucu tarafı
 * düştü". Kullanıcı için ikisi de beklemeyi gerektiriyor, ama metin farklı:
 * biri bağlantını kontrol et der, öteki birkaç saniye bekle.
 */
const GECICI_KALIPLAR: RegExp[] = [
  /gateway\s*time-?out/i,
  /bad\s*gateway/i,
  /service\s*unavailable/i,
  /internal\s*server\s*error/i,
  /\b(502|503|504)\b/,
  /statement\s*timeout|canceling\s*statement/i,
  /\btimed?\s*out\b|\btimeout\b/i,
  /upstream\s*(connect|request|error)/i,
  /ECONNRESET|ETIMEDOUT|EAI_AGAIN|ENOTFOUND/,
  /TimeoutException|SocketException|HandshakeException/,
];

/**
 * Kullanıcıya gösterilmemesi gereken "makine metni" kalıpları.
 *
 * ⚠ Bu bir KARA liste ve öyle olması bilinçli. Beyaz liste ("Türkçe mi?")
 * denenemez: kendi mesajlarımızın çoğu ASCII ("Ad zorunludur.", "Oturum
 * açık değil."), yani Türkçe karaktere bakan bir test onları da elerdi.
 * Kara liste yanılırsa hata GÜVENLİ tarafta olur: tanımadığımız bir metin
 * geçer, ekranda tuhaf ama okunabilir bir şey kalır — sessizce jenerikleşen
 * bir Türkçe mesajdan iyidir.
 */
const MAKINE_KALIPLARI: RegExp[] = [
  /^\s*[[{<]/, // JSON gövdesi ya da HTML sayfası
  /"(message|error|code|hint|details)"\s*:/i, // JSON parçası
  /^(Error|Exception|TypeError|RangeError):/,
  /(Postgrest|Auth(Api|Retryable|Session|Weak)?|Storage|Functions?(Http|Relay))[A-Za-z]*Exception/,
  /\bPGRST\d+\b/,
  /\b[0-9A-Z]{5}\b\s*(?:,|:)?\s*details:/i, // SQLSTATE dökümü
  /duplicate key|violates (unique|foreign key|check|not-null) constraint|null value in column/i,
  /\bJWT\b|jwt (expired|malformed)|invalid claim/i,
  /permission denied for (table|relation|schema|function)/i,
  /relation ".*" does not exist|column ".*" does not exist/i,
  /^\s*<!DOCTYPE/i,
  /Failed to fetch|Load failed|NetworkError|fetch failed|network request failed/i,
];

/** Metnin kendisi makine çıktısı mı (kullanıcıya gösterilemez mi)? */
export function isMachineMessage(message: string): boolean {
  const msg = message.trim();
  if (!msg) return true;
  return MAKINE_KALIPLARI.some((re) => re.test(msg));
}

/** Geçici sunucu arızası mı (bekleyip tekrar denemek işe yarar mı)? */
export function isTemporaryServerError(message: string): boolean {
  return GECICI_KALIPLAR.some((re) => re.test(message));
}

/** Hatadan ham mesaj metnini çıkarır (hata nesnesi olmayabilir). */
export function rawErrorMessage(err: unknown): string {
  if (err instanceof Error) return err.message ?? '';
  const m = (err as { message?: unknown } | null)?.message;
  if (typeof m === 'string') return m;
  if (err == null) return '';
  const s = String(err);
  return s === '[object Object]' ? '' : s;
}

export type FriendlyErrorOptions = {
  /** Makine metni yerine gösterilecek metin. Varsayılan: GENERIC_ERROR_NOTICE. */
  fallback?: string;
  /** Telemetriye yazılırken hangi yüzeyden geldiği (`etiket`). */
  surface?: string;
  /** Telemetriyi atla (testler ve zaten raporlayan çağıranlar için). */
  report?: boolean;
};

/**
 * Kullanıcıya gösterilecek metni döndürür — ham makine çıktısını asla
 * geçirmez, ham metni telemetriye yazar.
 *
 * Ağ (cihaz çevrimdışı) dalı BİLEREK burada değil: çağıranların kendi
 * bağlam metinleri var (`OFFLINE_MOVE_NOTICE`, `OFFLINE_MEANING_NOTICE`),
 * ve o ayrım `isNetworkError` ile çağıranda yapılıyor.
 */
export function friendlyErrorMessage(err: unknown, options: FriendlyErrorOptions = {}): string {
  const { fallback = GENERIC_ERROR_NOTICE, surface, report = true } = options;
  const raw = rawErrorMessage(err);

  // 1. Sunucunun kendi reddi — metin ne olursa olsun gösterilir.
  if (isServerRejection(err) && raw.trim()) return raw.trim();

  const bildir = () => {
    if (report) raporla?.(err, surface ? `hata-metni:${surface}` : 'hata-metni');
  };

  // 2. Geçici sunucu arızası. ⚠ Makine testinden ÖNCE: "Gateway Timeout"
  // düz metin olarak da gelebiliyor (JSON gövdesi olmadan) ve o hâliyle
  // hiçbir makine kalıbına takılmaz — vakanın kendisi bu.
  if (isTemporaryServerError(raw)) {
    bildir();
    return TEMPORARY_ERROR_NOTICE;
  }

  // 3. Makine metni — kullanıcıya Türkçe, telemetriye ham.
  if (isMachineMessage(raw)) {
    bildir();
    return fallback;
  }

  // 4. Kendi Türkçe mesajımız (ya da tanımadığımız ama okunabilir bir metin).
  return raw.trim();
}

File: mobile/app/lib/src/util/error_message.dart
// Kullanıcıya gösterilecek hata metninin son kapısı — web
// `src/utils/errorMessage.ts` portu.
//
// NEDEN (13 Eylül 2026): kullanıcı App Store için ekran kaydı çekerken giriş
// penceresinde `{"message":"Gateway Timeout"}` gördü — ham bir HTTP 504
// gövdesi, Türkçe bir uygulamada. İkinci denemede giriş çalıştı, yani arıza
// geçiciydi ama ekrandaki metin bunu söylemiyordu.
//
// **METİNLER VE KALIPLAR İKİ PLATFORMDA AYNI OLMAK ZORUNDA** — biri
// değişirse öteki de. `test/error_message_parity_test.dart` web dosyasını
// OKUYUP karşılaştırır (bulamazsa düşer).
//
// Dört dal, SIRASI davranışın parçası:
//   1. Sunucunun KENDİ reddi (SQLSTATE `P0001`) → olduğu gibi göster.
//      `submit_move` → "Sıra sende değil.", `accept_friend_invite` →
//      "Kendi linkinle arkadaş olamazsın." Bunlar KULLANICIYA gösterilmek
//      üzere yazılmış Türkçe metinler; koda bakılır, METNE değil.
//   2. Geçici sunucu arızası → "birkaç saniye sonra tekrar dene".
//      ⚠ 3'ten ÖNCE: "Gateway Timeout" düz metin olarak da geliyor ve o
//      hâliyle hiçbir makine kalıbına takılmaz — vaka tam olarak bu.
//   3. Makine metni → jenerik Türkçe cümle.
//   4. Geri kalan → olduğu gibi (kendi Türkçe doğrulama mesajlarımız).
//
// ⚠ Ağ (cihaz çevrimdışı) dalı BİLEREK burada değil: çağıranların kendi
// bağlam metinleri var (`kOfflineMoveNotice` gibi) ve o ayrım `isNetworkError`
// ile çağıranda yapılıyor.

/// Bilinmeyen/makine kaynaklı hata — kullanıcı ne yapacağını bilsin.
const kGenericErrorNotice = 'Bir sorun oluştu. Lütfen tekrar dene.';

/// Sunucu ayakta değil ya da zaman aşımına uğradı — GEÇİCİ olduğu söylenmeli.
///
/// Ayrı bir metin olmasının sebebi doğrudan vakanın kendisi: kullanıcı ikinci
/// denemede giriş yapabildi. "Bir sorun oluştu" doğru ama eksik; burada
/// yapılacak somut bir şey var ve o da beklemek.
const kTemporaryErrorNotice =
    'Sunucuya şu anda ulaşılamıyor. Birkaç saniye sonra tekrar dene.';

/// Ham metni nereye yazacağımız — DIŞARIDAN verilir (web'deki ile aynı
/// gerekçe: bu dosya saf kalsın, testler Supabase çekmesin).
typedef ErrorMessageReporter = void Function(Object? err, String context);
ErrorMessageReporter? _raporla;

/// Uygulama açılışında bağlanır; bağlanmazsa sessizce hiçbir şey yazılmaz.
void setErrorMessageReporter(ErrorMessageReporter? fn) => _raporla = fn;

final List<RegExp> _geciciKaliplar = [
  RegExp(r'gateway\s*time-?out', caseSensitive: false),
  RegExp(r'bad\s*gateway', caseSensitive: false),
  RegExp(r'service\s*unavailable', caseSensitive: false),
  RegExp(r'internal\s*server\s*error', caseSensitive: false),
  RegExp(r'\b(502|503|504)\b'),
  RegExp(r'statement\s*timeout|canceling\s*statement', caseSensitive: false),
  RegExp(r'\btimed?\s*out\b|\btimeout\b', caseSensitive: false),
  RegExp(r'upstream\s*(connect|request|error)', caseSensitive: false),
  RegExp(r'ECONNRESET|ETIMEDOUT|EAI_AGAIN|ENOTFOUND'),
  RegExp(r'TimeoutException|SocketException|HandshakeException'),
];

/// ⚠ KARA liste, bilinçli. Beyaz liste ("Türkçe mi?") denenemez: kendi
/// mesajlarımızın çoğu ASCII ("Ad zorunludur."), yani Türkçe karaktere bakan
/// bir test onları da elerdi. Kara liste yanılırsa hata GÜVENLİ tarafta olur.
final List<RegExp> _makineKaliplari = [
  RegExp(r'^\s*[\[{<]'),
  RegExp(r'"(message|error|code|hint|details)"\s*:', caseSensitive: false),
  RegExp(r'^(Error|Exception|TypeError|RangeError):'),
  RegExp(
      r'(Postgrest|Auth(Api|Retryable|Session|Weak)?|Storage|Functions?(Http|Relay))[A-Za-z]*Exception'),
  RegExp(r'\bPGRST\d+\b'),
  RegExp(r'\b[0-9A-Z]{5}\b\s*(?:,|:)?\s*details:', caseSensitive: false),
  RegExp(
      r'duplicate key|violates (unique|foreign key|check|not-null) constraint|null value in column',
      caseSensitive: false),
  RegExp(r'\bJWT\b|jwt (expired|malformed)|invalid claim',
      caseSensitive: false),
  RegExp(r'permission denied for (table|relation|schema|function)',
      caseSensitive: false),
  RegExp(r'relation ".*" does not exist|column ".*" does not exist',
      caseSensitive: false),
  RegExp(r'^\s*<!DOCTYPE', caseSensitive: false),
  // ⚠ Web'in kalıbının ÜST KÜMESİ: ilk beşi tarayıcının fetch metinleri,
  // gerisi Dart'ın (`dart:io`/`package:http`) taşıma metinleri — web'de hiç
  // oluşmazlar, bu yüzden web'e eklenmedi. Kalıp SAYISI parite testi için
  // aynı tutuldu (tek regex'e eklendi). 26 Eylül 2026, 1.1.1 cihaz turu:
  // uçak modunda Canlı oyun mesajı ekrana ham
  // "Failed host lookup: 'xvq….supabase.co'" bastı — `ClientException`in
  // `message` alanı sınıf adını TAŞIMIYOR, yani `SocketException` kalıbı
  // (`toString()`e bakan) onu göremiyordu.
  RegExp(
      r'Failed to fetch|Load failed|NetworkError|fetch failed|network request failed'
      r'|Failed host lookup|Connection (refused|reset|closed|timed out)'
      r'|Network is unreachable|No address associated|OS Error',
      caseSensitive: false),
];

/// Sunucunun bilerek yazdığı, kullanıcıya gösterilebilir ret (SQLSTATE P0001).
///
/// `ServerRejection`, `PostgrestException` ve `AuthException` üçü de `code`
/// alanı taşıyor ama ortak bir üst tipleri yok — bu yüzden `dynamic` okunuyor.
bool isServerRejection(Object? err) {
  if (err == null) return false;
  try {
    final dynamic d = err;
    return d.code == 'P0001';
  } catch (_) {
    return false;
  }
}

/// Metnin kendisi makine çıktısı mı (kullanıcıya gösterilemez mi)?
bool isMachineMessage(String message) {
  final msg = message.trim();
  if (msg.isEmpty) return true;
  return _makineKaliplari.any((re) => re.hasMatch(msg));
}

/// Geçici sunucu arızası mı (bekleyip tekrar denemek işe yarar mı)?
bool isTemporaryServerError(String message) =>
    _geciciKaliplar.any((re) => re.hasMatch(message));

/// Hatadan ham mesaj metnini çıkarır.
///
/// ⚠ `toString()` son çare: Dart'ta o çoğu zaman `PostgrestException(message:
/// …, code: …, details: …)` dökümünü verir — 3. dal onu zaten eler.
String rawErrorMessage(Object? err) {
  if (err == null) return '';
  try {
    final dynamic d = err;
    final m = d.message;
    if (m is String) return m;
  } catch (_) {
    // `message` alanı yok — aşağıdaki toString()'e düş.
  }
  return err.toString();
}

/// Kullanıcıya gösterilecek metni döndürür — ham makine çıktısını asla
/// geçirmez, ham metni telemetriye yazar.
String friendlyErrorMessage(
  Object? err, {
  String fallback = kGenericErrorNotice,
  String? surface,
  bool report = true,
}) {
  final raw = rawErrorMessage(err);

  // 1. Sunucunun kendi reddi — metin ne olursa olsun gösterilir.
  if (isServerRejection(err) && raw.trim().isNotEmpty) return raw.trim();

  void bildir() {
    final r = _raporla;
    if (report && r != null) {
      r(err, surface == null ? 'hata-metni' : 'hata-metni:$surface');
    }
  }

  // 2. Geçici sunucu arızası. ⚠ Makine testinden ÖNCE.
  if (isTemporaryServerError(raw)) {
    bildir();
    return kTemporaryErrorNotice;
  }

  // 3. Makine metni — kullanıcıya Türkçe, telemetriye ham.
  if (isMachineMessage(raw)) {
    bildir();
    return fallback;
  }

  // 4. Kendi Türkçe mesajımız.
  return raw.trim();
}
